test/service_test.cc

```diff
@@ -0,0 +1,458 @@
+      ASSERT_EQ(ServiceDataStructure::OK, ret);
+
+      // colect the texts posted
+      chirps_content_from_backend.push_back(chirp.text);
+    }
+    // Check if the contents are identical
+    EXPECT_EQ(chirps_content_after_delete, chirps_content_from_backend);
+  }
+}
+
+// This tests on `Follow` and `MonitorFrom` in `ServiceDataStructure`
+TEST_F(ServiceTestDataStructure, FollowAndMonitorTest) {
+  // Make each user follows the next user
+  for(size_t i = 0; i < kNumOfUsersPreset; ++i) {
+    // std::unique_ptr<ServiceDataStructure::UserSession>
+    auto session = service_data_structure_.UserLogin(user_list_[i]);
+    // Login should be successful
+    EXPECT_NE(nullptr, session);
+    // ServiceDataStructure::ReturnCodes
+    auto ret = session->Follow(user_list_[(i + 1) % kNumOfUsersPreset]);
+    // Following should be successful
+    EXPECT_EQ(ServiceDataStructure::OK, ret);
+    ret = session->Follow("non-existed");
+    // Should not follow a non-existed user
+    EXPECT_EQ(ServiceDataStructure::FOLLOWEE_NOT_FOUND, ret);
+  }
+
+  // Each user monitors their following users
+  for(size_t i = 0; i < kNumOfUsersPreset; ++i) {
+    // std::unique_ptr<ServiceDataStructure::UserSession>
+    auto session_user_followed = service_data_structure_.UserLogin(user_list_[(i + 1) % kNumOfUsersPreset]);
+    // Login as the followed user should be successful
+    EXPECT_NE(nullptr, session_user_followed);
+
+    // Post some dont-care chirps from the followed user
+    for(size_t j = 0; j < 5; ++j) {
+      auto ret = session_user_followed->PostChirp(kShortText, nullptr);
+      // Posting should be successful
+      EXPECT_EQ(ServiceDataStructure::OK, ret);
+    }
+
+    // Timestamp the current time and back it up
+    struct timeval now;
+    gettimeofday(&now, nullptr);
+    struct timeval backup_now = now;
+
+    // sleep a little while to ensure that the time has passed at least 1 usec.
+    std::this_thread::sleep_for(std::chrono::microseconds(1));
+
+    // Collect the chirp ids after the above timestamp
+    std::set<uint64_t> chirp_collector;
+    for(size_t j = 0; j < 5; ++j) {
+      uint64_t chirp_id;
+      auto ret = session_user_followed->PostChirp(kShortText, &chirp_id);
+      // Posting should be successful
+      EXPECT_EQ(ServiceDataStructure::OK, ret);
+      chirp_collector.insert(chirp_id);
+    }
+
+    // Test Monitoring from
+    // std::unique_ptr<ServiceDataStructure::UserSession>
+    auto session = service_data_structure_.UserLogin(user_list_[i]);
+    // Login should be successful
+    EXPECT_NE(nullptr, session);
+    auto monitor_result = session->MonitorFrom(&now);
+    // `now` should be modified by the `Monitor` function
+    // so it should be different from the one I have backed up
+    EXPECT_TRUE(timercmp(&backup_now, &now, !=));
+    // Check if the contents are identical
+    EXPECT_EQ(chirp_collector, monitor_result);
+  }
+}
+
+// This test cases on the Service Server to check whether their
```

Comment:
I'm not sure if you need to test this connection. I believe this is testing the connection between the service_server and the actual structure computing the service layer requests. Ideally these are just single method calls, so testing those methods are sufficient.

Reply:
This is the thing that gave me some trouble since my methods behind the service-layer server do not share the same interfaces with the server.

Reply:
Makes sense